File: GDAY/simulations/quasi_equil_annual_simulations.py
#!/usr/bin/env python

""" Quasi-equilibrium Simulations

- Model simulation: using hypothetical annual met forcing data and hypothetical parameters
                 to run sensitivity tests.

"""

import os
import shutil
import sys
import subprocess
USER = os.getlogin()
# get the parent directory of the current working directory
from os.path import dirname, abspath
d = dirname(dirname(abspath(__file__)))

sys.path.append(os.path.join(d, "code/scripts"))

import adjust_gday_param_file as ad

__author__  = "Martin De Kauwe"
__version__ = "1.0 (05.08.2016)"
__email__   = "mdekauwe@gmail.com"

def main(site, treatment):

    GDAY_SPIN = "./gday -s -p "
    GDAY = "./gday -p "

    # dir names
    base_dir = os.path.dirname(os.getcwd())
    param_dir = os.path.join(d, "params")
    met_dir = os.path.join(d, "met_data")
    run_dir = os.path.join(d, "outputs")
    
    
    shutil.copy(os.path.join(param_dir, "%s_model_spunup.cfg" % (site)),
                os.path.join(param_dir, "%s_model_transient.cfg" % (site)))

    itag = "%s_model_transient" % (site)
    otag = "%s_model_simulation_%s" % (site, treatment)
    mtag = "%s_met_forcing_transient_%s.csv" % (site, treatment)
    out_fn = "%s_transient_%s.csv" % (site, treatment.upper())
    out_param_fname = os.path.join(param_dir, otag + ".cfg")
    cfg_fname = os.path.join(param_dir, itag + ".cfg")
    met_fname = os.path.join(met_dir, mtag)
    out_fname = os.path.join(run_dir, out_fn)

    replace_dict = {
                     # files
                     "out_param_fname": "%s" % (out_param_fname),
                     "cfg_fname": "%s" % (cfg_fname),
                     "met_fname": "%s" % (met_fname),
                     "out_fname": "%s" % (out_fname),


                     # control
                     "print_options": "annual",

                    }
    ad.adjust_param_file(cfg_fname, replace_dict)
    os.system(GDAY + cfg_fname)


    # add this directory to python search path so we can find the scripts!
    sys.path.append(os.path.join(d, "code/scripts"))
    import translate_GDAY_output_to_NCEAS_format as tr
    # tr.translate_output(out_fname, met_fname)


if __name__ == "__main__":



    site = "Quasi_equil"

    # Ambient
    main(site, treatment="co2_amb")

    # Elevated
    main(site, treatment="co2_ele")

File: GDAY/simulations/Run5/quasi_equil_annual_simulations.R
#### Quasi-equilibrium Simulations

#### Model simulation: using hypothetical annual met forcing data and hypothetical parameters
#### to run sensitivity tests.


#### Author: Mingkai Jiang (m.jiang@westernsydney.edu.au)
#### Date Created: Mar-30-2017
##########################################################################################

############################# Sourcing outside functions #################################

### get the parent directory of the current working directory
d <- dirname(dirname(getwd()))   

### sourcing the adjust_gday_param_file.py code    # Note: need to change to R
script_path <- paste0(d, "/code/scripts")
source(paste0(script_path, "/adjust_gday_param_file.R"))

################################ Main functions #########################################
Run_GDAY_transient <- function(site, treatment) {
    
    #### Setting executive commands
    GDAY_SPIN <- "./gday -s -p"
    GDAY <- "./gday -p"
    
    
    #### setting directory names
    base_dir <- getwd()
    param_dir <- paste0(d, "/params/Run5")
    met_dir <- paste0(d, "/met_data")
    run_dir <- paste0(d, "/outputs/Run5")
    
    
    #### setting up the output file names and locations
    itag <- paste0(site, "_model_transient")
    otag <- paste0(site, "_model_simulation_", treatment)
    mtag = paste0(site, "_met_forcing_transient_", treatment, ".csv")
    out_fn <- paste0(site, "_transient_", toupper(treatment), ".csv")   
    out_param_fname <- paste0(param_dir, "/", otag, ".cfg")
    cfg_fname <- paste0(param_dir, "/", itag, ".cfg")
    met_fname <- paste0(met_dir, "/", mtag)
    out_fname <- paste0(run_dir, "/", out_fn)
    
    #### Copy and paste the initial parameter cfg file
    sys_com1 <- paste0("cp ", param_dir, "/", site, "_model_spunup.cfg ",
                       cfg_fname)
    system(paste(sys_com1))
 
    
    #### set up the replacement list
    replace_dict <- c(
        ############## FILES ############
        "out_param_fname", out_param_fname,
        "cfg_fname", cfg_fname,
        "met_fname", met_fname,
        "out_fname", out_fname,
        ############## CONTROL ############
        "print_options", "annual")
    
    #### make a df out from replacement dictionary
    rDF <- make_df(replace_dict)
    
    #### call function to conduct the parameter replacement
    adjust_gday_params(cfg_fname, rDF)
    
    #### Run the spin up model
    system(paste0(GDAY, " ", cfg_fname))

}


################################ Program #########################################
site = "Quasi_equil"

# Ambient
Run_GDAY_transient(site, treatment="co2_amb")

# Elevated
Run_GDAY_transient(site, treatment="co2_ele")
